File: solutions/479/main.py
# TLE at n = 4 
class Solution:
    def isPalNum(self, num):
        num = [int(x) for x in str(num)]
        i, j = 0, len(num) - 1 
        while i < j: 
            if num[i] != num[j]: 
                return False 
            i += 1 
            j -= 1
        return True 
    def largestPalindrome(self, n: int) -> int:
        max_res = 0 
        for i in range(10**n - 1, 10**(n-1)-1, -1): 
            for j in range(10**n - 1, 10**(n-1)-1, -1): 
                res = i * j 
                if self.isPalNum(res):
                    max_res = max(res, max_res) 
        return max_res % 1337 

if __name__ == "__main__": 
    s = Solution()

    n = 2 
    result = s.largestPalindrome(n)
    assert result ==  9009 % 1337

    n = 1
    result = s.largestPalindrome(n)
    assert result ==  9 % 1337

File: solutions/479/main.rs
use std::collections::BinaryHeap;
impl Solution {
    pub fn is_pal(n: i64) -> bool {
        let mut num = n; 
        let mut rev = 0; 
        let mut p = (n as f64).log10() as u32; 
        while num > 0 {
            rev += 10i64.pow(p) * (num % 10); 
            num /= 10; 
            p = if p >= 1 {p-1} else {0}; 
        } 
        return rev == n;
    }
    pub fn largest_palindrome(n: i32) -> i32 {
        let n = n as u32; 
        let mut pq = BinaryHeap::new(); 
        for i in (10i64.pow(n-1)-1..10i64.pow(n)-1).rev() {
            pq.push((i * (10i64.pow(n)-1), i, 10i64.pow(n)-1));
        }
        loop {
            let (res, i, j) = pq.pop().unwrap(); 
            if Solution::is_pal(res) {
                return (res % 1337) as i32; 
            } 
            pq.push((i * (j-1), i, j-1)); 
        }
    }
}

struct Solution {}

fn main() {
    // assert_eq!(Solution::largest_palindrome(1), 9); 
    assert_eq!(Solution::largest_palindrome(7), 877); 
}